### PR TITLE
ML UI + Live Joints Feedback

### PR DESCRIPTION
**Pull request by Siupahc**
# Summary 
This PR introduces the ML coach, which allows users to get feedback on their workout. In a friendly manner. The UI is now much cleaner than before

# Introduction to the ML Coach:
The ML coach implementation reuses the backend added from the PR #157 and a UI is added on top of this one. There are two major updates: 
1- Live joints feedback : The user gets feedback in real time by displaying his joints on top of his body. Joints that require attention are highlighted so that the user can pay attention to them.  
2- Coach feedback:  The UI is now much better, and more user-friendly. It also gives a grade according to the quality of movements.

## Minor features
The Video Screen has been rebranded to the Learning Screen, since that’s the entry point for the whole ML Coach feature.
The skeleton displayed in the live feedback can be dragged so that the user (or at least the one who’s filming him) can freely analyze the joints. 
  
### Who did what
The UI part was done by Louis
The live joints feedback part by Shayan
Why did we both work on the same branch? Good question, but we couldn’t solve the issue so there is one PR for both of us

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/screen/CoachCaptureScreenTest.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import com.android.sample.model.camera.CameraViewModel
+import com.android.sample.ui.mlFeedback.CoachCaptureScreen
+import com.android.sample.ui.navigation.NavigationActions
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.Mockito.mock
+
+class CoachCaptureScreenTest {
+  private lateinit var mockCameraViewModel: CameraViewModel
```
Comment (didi2101) [inline app/src/androidTest/java/com/android/sample/screen/CoachCaptureScreenTest.kt:18]: Why mock the CameraViewModel if not to use it ?

File: app/src/androidTest/java/com/android/sample/screen/CoachFeedBackScreenTest.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    composeTestRule.onNodeWithTag("rankText").assertIsDisplayed().assertTextEquals("S")
+    composeTestRule.onNodeWithTag("exerciseCard").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("exerciseName").assertIsDisplayed().assertTextEquals("Plank")
+    composeTestRule
+        .onNodeWithTag("exerciseDuration")
+        .assertIsDisplayed()
+        .assertTextContains("10 s", substring = true)
+    composeTestRule.onNodeWithTag("doneButton").assertIsDisplayed().assertHasClickAction()
+  }
+}
```
Comment (didi2101) [inline app/src/androidTest/java/com/android/sample/screen/CoachFeedBackScreenTest.kt:73]: Good tests !

Maybe add little comment or Kdoc.

File: app/src/main/java/com/android/sample/mlUtils/CoachFeedBack.kt [hunk 3]
```diff
@@ -19,18 +19,31 @@ data class CoachFeedback(
...
    */
   override fun toString(): String {
+    val stringBuilder: StringBuilder = StringBuilder()
+    // stringBuilder.append(exerciseCriterion.name).append("\n")
+    commentSet
+        .filter { it.rate >= 0.1F }
+        .forEach { comment -> stringBuilder.append(comment.comment).append("\n") }
+
+    //    stringBuilder.append(
+    //        "${feedbackUnit.valuePrefix}: $feedbackValue ${feedbackUnit.stringRepresentation}\n")
```
Comment (didi2101) [inline app/src/main/java/com/android/sample/mlUtils/CoachFeedBack.kt:37]: Just delete this useless commented code.

File: app/src/main/java/com/android/sample/mlUtils/CoachFeedBack.kt [hunk 4]
```diff
@@ -51,3 +64,23 @@ data class JointFeedback(val comment: String = "", val rate: Float = 0F) {
...
+fun rateToRank(rate: Float): FeedbackRank {
+  return when {
+    rate >= 0.9 -> FeedbackRank.S
+    rate >= 0.8 -> FeedbackRank.A
+    rate >= 0.7 -> FeedbackRank.B
+    rate >= 0.6 -> FeedbackRank.C
+    rate > 0.1 -> FeedbackRank.D
+    else -> FeedbackRank.X
+  }
+}
```
Comment (didi2101) [inline app/src/main/java/com/android/sample/mlUtils/CoachFeedBack.kt:86]: Nice code ! 

Very clear.

File: app/src/main/java/com/android/sample/mlUtils/PoseDetectionJoints.kt [hunk 5]
```diff
@@ -51,5 +72,118 @@ class PoseDetectionJoints {
...
+        jointColor: Color = DarkBlue,
+        jointStroke: Float = 15f,
+        lineColor: Color = LightBlue,
+        lineStroke: Float = 8f,
+        wrongJointColor: Color = Color.Yellow,
+        wrongJointStroke: Float = 15f,
+        wronglineColor: Color = Color.Red,
+        wronglineStroke: Float = 8f,
+        modifier: Modifier = Modifier
+    ) {
```
Comment (didi2101) [inline app/src/main/java/com/android/sample/mlUtils/PoseDetectionJoints.kt:107]: Add Kdoc for DrawBody here.

File: app/src/main/java/com/android/sample/mlUtils/exercisesCriterions/AngleCriterionComments.kt [hunk 6]
```diff
@@ -1,39 +1,167 @@
...
+      "Tighten more your right leg.",
+      focusedJoints = listOf(PoseDetectionJoints.RIGHT_OPPHIP_HIP_KNEE)),
+  LEFT_OPPHIP_HIP_KNEE_NOT_CLOSED(
+      "Tighten more your left leg.",
+      focusedJoints = listOf(PoseDetectionJoints.LEFT_OPPHIP_HIP_KNEE)),
+  BOTH_OPPHIP_HIP_KNEE_NOT_CLOSED(
+      "Tighten more your legs",
+      focusedJoints =
+          listOf(
+              PoseDetectionJoints.LEFT_OPPHIP_HIP_KNEE, PoseDetectionJoints.RIGHT_OPPHIP_HIP_KNEE))
```
Comment (didi2101) [inline app/src/main/java/com/android/sample/mlUtils/exercisesCriterions/AngleCriterionComments.kt:166]: Put a period at the end of each sentence, or not, but don't mix the two.

I'd vote for a period.

**Review by didi2101 — CHANGES_REQUESTED** [review]
Very Good feature overall !

This will win us the future battle of the app.

Just apply the changes needed, add this Kdoc !

File: app/src/main/java/com/android/sample/ui/composables/SaveButton.kt [hunk 7]
```diff
@@ -44,10 +44,10 @@ fun SaveButton(onSaveClick: () -> Unit, testTag: String) {
       contentPadding = PaddingValues()) {
         Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
           Row(verticalAlignment = Alignment.CenterVertically) {
-            Icon(imageVector = Icons.Default.Check, contentDescription = "Save", tint = Color.White)
+            Icon(imageVector = Icons.Default.Check, contentDescription = text, tint = Color.White)
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/composables/SaveButton.kt:47]: lnPSQzZg

File: app/src/main/java/com/android/sample/ui/mlFeedback/CoachFeedbackScreen.kt [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+private fun getColorForRank(rank: FeedbackRank): Color {
+  return when (rank) {
+    FeedbackRank.S -> YogaTag
+    FeedbackRank.A -> Green
+    FeedbackRank.B -> RunningTag
+    FeedbackRank.C -> Yellow
+    FeedbackRank.D -> Red
+    FeedbackRank.X -> Black
+  }
+}
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/mlFeedback/CoachFeedbackScreen.kt:279]: Will do

File: app/src/main/java/com/android/sample/ui/video/VideoLibraryScreen.kt [hunk 9]
```diff
@@ -61,6 +68,46 @@ fun VideoLibraryScreen(navigationActions: NavigationActions, videoViewModel: Vid
...
+          else -> {
+            Box(modifier = Modifier.padding(16.dp)) {
+              Column(horizontalAlignment = Alignment.CenterHorizontally) {
+                Box(
+                    modifier =
+                        Modifier.width(
+                                175.dp) // Set a fixed width for the text box to prevent everything
+                            // from moving
+                            .background(
+                                Color.Black.copy(alpha = 0.7f), shape = RoundedCornerShape(8.dp))
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/video/VideoLibraryScreen.kt:83]: ^^

File: app/src/main/java/com/android/sample/ui/mlFeedback/CoachCaptureScreen.kt [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+      onDismissRequest = onDismissRequest,
+      modifier = Modifier.testTag("infoDialogue"),
+  ) {
+    Card {
+      Column(modifier = Modifier.padding(16.dp)) {
+        Text(
+            text = "How to use the Coach",
+            style = MaterialTheme.typography.titleMedium,
+            modifier = Modifier.padding(bottom = 8.dp))
+        Text("Press record and start your exercise\n")
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/mlFeedback/CoachCaptureScreen.kt]: Yes

File: app/src/main/java/com/android/sample/ui/mlFeedback/CoachCaptureScreen.kt [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        Text("Your whole body has to be visible in the camera frame.\n")
+        Text(
+            "To ensure proper feedback, please be perpendicular to the camera. For instance if you are doing a plank, you should no be facing the camera.\n")
+        Text("Coach will highlight mistakes in real time.")
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/mlFeedback/CoachCaptureScreen.kt:268]: Added.

File: app/src/main/java/com/android/sample/ui/composables/TalkingCoach.kt [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        painter = painterResource(id = R.drawable.endurai_coach),
+        contentDescription = "Coach",
+        modifier =
+            Modifier.size(size)
+                .clip(CircleShape)
+                .shadow(8.dp, CircleShape)
+                .background(BlueGradient, CircleShape)
+                .testTag("coachImage"))
+  }
+}
```
[the remark this answers is not in the document]
Comment (Siupahc) [reply app/src/main/java/com/android/sample/ui/composables/TalkingCoach.kt:62]: Added

File: app/src/androidTest/java/com/android/sample/screen/CoachCaptureScreenTest.kt (hunk 1 above)
Comment (Siupahc) [reply app/src/androidTest/java/com/android/sample/screen/CoachCaptureScreenTest.kt]: Good question. Changed.

File: app/src/main/java/com/android/sample/ui/composables/SaveButton.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (didi2101) [reply app/src/main/java/com/android/sample/ui/composables/SaveButton.kt:47]: Chef ?

**Review by didi2101 — APPROVED** [review]
Modifications are done.

Approved.

**Pull request merged**